connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamingQueryCache.scala

```diff
@@ -0,0 +1,203 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connect.service
+
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+import javax.annotation.concurrent.GuardedBy
+
+import scala.collection.mutable
+import scala.concurrent.duration.Duration
+import scala.concurrent.duration.DurationInt
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.streaming.StreamingQuery
+import org.apache.spark.util.Clock
+import org.apache.spark.util.SystemClock
+
+/**
+ * Caches Spark-Connect streaming query references and the sessions. When a query is stopped (i.e.
+ * no longer active), it is cached for 1 hour so that it is accessible from the client side. It
+ * runs a background thread to run a periodic task that does the following:
+ *   - Check the status of the queries, and drops those that expired (1 hour after being stopped).
+ *   - Keep the associated session active by invoking supplied function `sessionKeepAliveFn`.
+ *
+ * This class helps with supporting following semantics for streaming query sessions:
+ *   - Keep the session and session mapping at connect server alive as long as a streaming query
+ *     is active. Even if the client side has disconnected.
+ *     - This matches how streaming queries behave in Spark. The queries continue to run if
+ *       notebook or job session is lost.
+ *   - Once a query is stopped, the reference and mappings are maintained for 1 hour and will be
+ *     accessible from the client. This allows time for client to fetch status.
+ *     - During this time if the query is restarted (i.e. has a new run id), the reference to
+ *       previous run is dropped. As a result logical query has only the most recent query
+ *       reference cached. This policy can be revisited to cache multiple runs for a query.
+ *   - Note that these semantics are evolving and might change before being finalized in Connect.
+ *   - Future improvements:
+ *     - Provide an API for a users to access session even after they lose session id.
+ *       - Once a user is properly authenticated, the API could return list of sessions that are
+ *         still alive in the connect server for that user.
```

Comment:
Added full semantics to Scaladoc. Let's move the discussion about them here. 

@juliuszsompolski the [points you raised](https://github.com/apache/spark/pull/40937#pullrequestreview-1399458578) important. PTAL above Scaladoc. Short answer: session will live as long as the query is active. Same as current Spark.

Reply:
Met with Julek about these semantics and larger life cycle. We agreed these make sense and will allow improvements to overall session management. 
E.g. we likely need to track batch queries too similar to streaming queries. We will likely fold this into new session manager. In addition, such improvements will not affect client API or protocol.

Reply:
Yes. I recently started working on designing more session management in general. We need to track all active running queries as well, to be able to e.g. interrupt/cancel it, or to re-attach to them in case client loses grpc stream connection, and also to avoid removing a session that still has queries that are running (which currently can happen if a client is just waiting in a ExecutePlan that is running for > 1 hour; session will get evicted based on 1 hour idleness...).
As result of this, this will likely be folded into the more generic session manager.

I'd remove the Future improvements comment from here for now, as the direction may in the end depend on the proposed design in general. That should come soon.

Reply:
Agreed. Removed 'Future improvements' section. 
I will update here if we can access a streaming query across different sessions/notebooks in classic spark.